src/python/pants/engine/native.py

```diff
@@ -387,13 +187,72 @@ def _replace_file(path, content):
     f.write(content)
 
 
-def _initialize_externs(ffi):
-  """Initializes extern callbacks given a CFFI handle."""
+class ExternSignature(datatype([
+    ('return_type', str),
+    ('method_name', str),
+    ('arg_types', tuple),
+])):
+  """A type signature for a python-defined FFI function."""
+
+  def pretty_print(self):
+    return '  {ret}\t{name}({args});'.format(
+      ret=self.return_type,
+      name=self.method_name,
+      args=', '.join(self.arg_types))
+
+
+def extern_decl(return_type, arg_types):
+  """A decorator for methods corresponding to extern functions. All types should be strings.
+
+  The FFISpecification class is able to automatically convert these into method declarations for
+  cffi.
+  """
+  def wrapper(func):
+    signature = ExternSignature(
+      return_type=str(return_type),
+      method_name=str(func.__name__),
+      arg_types=tuple(arg_types))
+    func.extern_signature = signature
+    return func
+  return wrapper
+
+
+class FFISpecification(object):
```

Comment:
Could classes that are not meant for consumption outside this file be prefixed with `_`? `_FFISpecification` and etc.

Reply:
Yes! Will do.

Reply:
Done!